File: bloomberg/98.validate-binary-search-tree.cpp
/*
 * [98] Validate Binary Search Tree
 *
 * https://leetcode.com/problems/validate-binary-search-tree
 *
 * Medium (23.09%)
 * Total Accepted:    
 * Total Submissions: 
 * Testcase Example:  '[2,1,3]'
 *
 * Can you solve this problem? 🤔
 */
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode(int x) : val(x), left(NULL), right(NULL) {}
 * };
 */
class Solution {
public:
    bool isValidBST(TreeNode* root) {
        if (!root) {
            //not a BST, let a lone a valid one
            return false;
        }
        int min = root->val;
        int max = root->val;
        return helper(root, min , max);
    }
    bool helper(TreeNode *root, int &min, int &max) {
        if (!root) {
            //return true in helper to not mess up
            return true;
        }
        //min and max values
        int leftMin, leftMax;
        int rightMin, rightMax;
        bool left = helper(root->left, leftMin, leftMax);
        bool right = helper(root->right, rightMin, rightMax);
        if (left == false || right == false) return false;
        if (root->left &&leftMax >= root->val) return false;
        if (root->right && rightMin <= root->val) return false;
        min = (root->left)? leftMin : root->val;
        max = (root->right)? rightMax : root->val;
        return true;

    }
};

File: advanced/facebook/second/98.validate-binary-search-tree.java
/*
 * [98] Validate Binary Search Tree
 *
 * https://leetcode.com/problems/validate-binary-search-tree
 *
 * algorithms
 * Medium (23.42%)
 * Total Accepted:    180.8K
 * Total Submissions: 771.8K
 * Testcase Example:  '[2,1,3]'
 *
 * 
 * Given a binary tree, determine if it is a valid binary search tree (BST).
 * 
 * 
 * 
 * Assume a BST is defined as follows:
 * 
 * The left subtree of a node contains only nodes with keys less than the
 * node's key.
 * The right subtree of a node contains only nodes with keys greater than the
 * node's key.
 * Both the left and right subtrees must also be binary search trees.
 * 
 * 
 * 
 * Example 1:
 * 
 * ⁠   2
 * ⁠  / \
 * ⁠ 1   3
 * 
 * Binary tree [2,1,3], return true.
 * 
 * 
 * Example 2:
 * 
 * ⁠   1
 * ⁠  / \
 * ⁠ 2   3
 * 
 * Binary tree [1,2,3], return false.
 * 
 */
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode(int x) { val = x; }
 * }
 */
class Solution {
    public boolean isValidBST(TreeNode root) {
        int[] stat =  new int[2];
        return helper(root, stat);
    }

    //util function
    private boolean helper(TreeNode root, int[] stat) {
    	if(root == null) {
    		return true;
    	}

    	int[] leftStat = new int[2];
    	int[] rightStat = new int[2];
    	boolean left = helper(root.left, leftStat);
    	boolean right = helper(root.right,rightStat);
    	if (left == false || right == false) return false;
    	if (root.left != null && leftStat[1] >= root.val) return false;
    	if (root.right != null && rightStat[0] <= root.val) return false;
    	stat[0] = (root.left == null) ? root.val : leftStat[0];
    	stat[1] = (root.right == null) ? root.val : rightStat[1];
    	return true;
    }
}
